<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-и-подготовка-данных" data-toc-modified-id="Загрузка-и-подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка и подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Признаки" data-toc-modified-id="Признаки-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Признаки</a></span></li><li><span><a href="#Целевой-признак" data-toc-modified-id="Целевой-признак-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Целевой признак</a></span></li></ul></li><li><span><a href="#Обучение-и-проверка-модели" data-toc-modified-id="Обучение-и-проверка-модели-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение и проверка модели</a></span><ul class="toc-item"><li><span><a href="#Регион-1" data-toc-modified-id="Регион-1-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Регион 1</a></span></li><li><span><a href="#Регион-2" data-toc-modified-id="Регион-2-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Регион 2</a></span></li><li><span><a href="#Регион-3" data-toc-modified-id="Регион-3-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Регион 3</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Подготовка-к-расчёту-прибыли" data-toc-modified-id="Подготовка-к-расчёту-прибыли-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Подготовка к расчёту прибыли</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Расчёт-прибыли-и-рисков" data-toc-modified-id="Расчёт-прибыли-и-рисков-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Расчёт прибыли и рисков</a></span><ul class="toc-item"><li><span><a href="#Функция-для-расчёта-прибыли-по-выбранным-скважинам-и-предсказаниям-модели:" data-toc-modified-id="Функция-для-расчёта-прибыли-по-выбранным-скважинам-и-предсказаниям-модели:-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Функция для расчёта прибыли по выбранным скважинам и предсказаниям модели:</a></span></li><li><span><a href="#Функция-для-расчета-рисков-и-прибыли-для-каждого-региона:" data-toc-modified-id="Функция-для-расчета-рисков-и-прибыли-для-каждого-региона:-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Функция для расчета рисков и прибыли для каждого региона:</a></span></li><li><span><a href="#Регион-1" data-toc-modified-id="Регион-1-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Регион 1</a></span></li><li><span><a href="#Регион-2" data-toc-modified-id="Регион-2-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Регион 2</a></span></li><li><span><a href="#Регион-3" data-toc-modified-id="Регион-3-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Регион 3</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.6"><span class="toc-item-num">4.6&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Описание проекта:

Нужно решить, где бурить новую скважину.

Вам предоставлены пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. Постройте модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализируйте возможную прибыль и риски техникой Bootstrap.


**Шаги для выбора локации:**

- В избранном регионе ищут месторождения, для каждого определяют значения признаков;
- Строят модель и оценивают объём запасов;
- Выбирают месторождения с самым высокими оценками значений. Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

# Описание данных

Данные геологоразведки трёх регионов находятся в файлах:
- geo_data_0.csv. 
- geo_data_1.csv. 
- geo_data_2.csv.

id — уникальный идентификатор скважины;

f0, f1, f2 — три признака точек (неважно, что они означают, но сами признаки значимы);

product — объём запасов в скважине (тыс. баррелей).

## Загрузка и подготовка данных

### Признаки

- `id` — уникальный идентификатор скважины;
- `f0, f1, f2` — три признака точек (неважно, что они означают, но сами признаки значимы).

### Целевой признак

- `product` — объём запасов в скважине (тыс. баррелей).

In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

from sklearn.model_selection import GridSearchCV


In [2]:
def df_info(file_path):
    df = pd.read_csv(file_path)
    display(df.head())
    display(df.info())
    display(df.describe())
    print('Количество дубликатов:', df.duplicated().sum())
    return df

In [3]:
geo_df_0 = df_info('geo_data_0.csv')

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.343769      16.003790     185.364347

Количество дубликатов: 0


In [4]:
geo_df_1 = df_info('geo_data_1.csv')

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%         1.153055      -4.813172       2.011479      57.085625
75%         8.621015      -1.332816       3.999904     107.813044
max        29.421755      18.734063       5.019721     137.945408

Количество дубликатов: 0


In [5]:
geo_df_2 = df_info('geo_data_2.csv')

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.002023      -0.002081       2.495128      95.000000
std         1.732045       1.730417       3.473445      44.749921
min        -8.760004      -7.084020     -11.970335       0.000000
25%        -1.162288      -1.174820       0.130359      59.450441
50%         0.009424      -0.009482       2.484236      94.925613
75%         1.158535       1.163678       4.858794     130.595027
max         7.238262       7.844801      16.739402     190.029838

Количество дубликатов: 0


Пропусков и дубликатов нет, предобработка не требуется.

In [6]:
#Определим признаки и целовой признак:
features = ['f0', 'f1', 'f2']
target = ['product']
features_0 = geo_df_0[features]
target_0 = geo_df_0[target]
features_1 = geo_df_1[features]
target_1 = geo_df_1[target]
features_2 = geo_df_2[features]
target_2 = geo_df_2[target]

## Обучение и проверка модели

Для каждого региона разделим данные на обучающую и валидационную выборки и выведем средний запас предсказанного сырья и RMSE модели.

In [7]:
def model_fitting(features, target):
    features_train, features_valid, target_train, target_valid = train_test_split(features, target, 
                                                                                  test_size=0.25, random_state=0)
    print('Размер тренировочной выборки:', features_train.shape)
    print('Размер валидационной выборки:', features_valid.shape)
    model = LinearRegression()
    model.fit(features_train, target_train)
    predictions = model.predict(features_valid)
    predictions_mean = predictions.mean()
    rmse = mean_squared_error(target_valid, predictions) ** 0.5
    print('Cредний запас предсказанного сырья: {:.4f}'.format(predictions_mean))
    print('RMSE модели: {:.4f}'.format(rmse))
    
    return target_valid, predictions

### Регион 1

In [8]:
target_valid_0, predictions_0 = model_fitting(features_0, target_0)

Размер тренировочной выборки: (75000, 3)
Размер валидационной выборки: (25000, 3)
Cредний запас предсказанного сырья: 92.2714
RMSE модели: 37.4810


### Регион 2

In [9]:
target_valid_1, predictions_1 = model_fitting(features_1, target_1)

Размер тренировочной выборки: (75000, 3)
Размер валидационной выборки: (25000, 3)
Cредний запас предсказанного сырья: 69.1516
RMSE модели: 0.8873


### Регион 3

In [10]:
target_valid_2, predictions_2 = model_fitting(features_2, target_2)

Размер тренировочной выборки: (75000, 3)
Размер валидационной выборки: (25000, 3)
Cредний запас предсказанного сырья: 94.7075
RMSE модели: 40.3129


### Вывод
Объем предсказанного сырья достаточно высокий для первого и третьего региона (92.3 и 94.7 соответственно), однако и RMSE этих моделей тоже высок. Объем предсказанного сырья для второго региона - 69, но и показатель RMSE тоже низкий.

## Подготовка к расчёту прибыли

Рассчитаем достаточный объём сырья для безубыточной разработки новой скважины и сравним полученный объём сырья со средним запасом в каждом регионе.

* `BUDGET` - бюджет на разработку скважин в регионе
* `TOP_WELLS` - всего скважин
* `TOTAL_WELLS` - кол-во выбранных скважин
* `UNIT_INCOME` - доход с каждой единицы продукта

In [11]:
BUDGET = 10**10
TOP_WELLS = 200
TOTAL_WELLS = 500
UNIT_INCOME = 450000

In [12]:
min_product = BUDGET / (TOP_WELLS * UNIT_INCOME)
print('Достаточный объём сырья для безубыточной разработки новой скважины: {:.0f}'.format(min_product), 
      'тыс. баррелей')

Достаточный объём сырья для безубыточной разработки новой скважины: 111 тыс. баррелей


### Вывод
Достаточный объём сырья для безубыточной разработки новой скважины составляет 111 тыс. баррелей. Данный объем превышает средний запас в каждом регионе.

## Расчёт прибыли и рисков 

### Функция для расчёта прибыли по выбранным скважинам и предсказаниям модели:
* Выберем скважины с максимальными значениями предсказаний
* Просуммируем целевое значение объёма сырья, соответствующее этим предсказаниям
* Рассчитаем прибыль для полученного объёма сырья

In [13]:
target_valid_0 = pd.Series(target_valid_0['product']).reset_index(drop=True)
target_valid_1 = pd.Series(target_valid_1['product']).reset_index(drop=True)
target_valid_2 = pd.Series(target_valid_2['product']).reset_index(drop=True)

In [14]:
predictions_0 = pd.Series(predictions_0[:, 0])
predictions_1 = pd.Series(predictions_1[:, 0])
predictions_2 = pd.Series(predictions_2[:, 0])

In [15]:
def predict_income(target, predictions):
    top_predictions = predictions.sort_values(ascending=False)
    top_target = target[top_predictions.index][:TOP_WELLS]
    income = top_target.sum() * UNIT_INCOME
    return income - BUDGET

In [16]:
print('Прибыль с лучших 200 скважин в первом регионе: {:.0f}'.format(predict_income(target_valid_0, 
                                                            predictions_0)), ' руб.')
print('Прибыль с лучших 200 скважин во втором регионе: {:.0f}'.format(predict_income(target_valid_1, 
                                                            predictions_1)), 'руб.')
print('Прибыль с лучших 200 скважин в третьем регионе: {:.0f}'.format(predict_income(target_valid_2, 
                                                            predictions_2)), 'руб.')

Прибыль с лучших 200 скважин в первом регионе: 3363408080  руб.
Прибыль с лучших 200 скважин во втором регионе: 2415086697 руб.
Прибыль с лучших 200 скважин в третьем регионе: 2623878684 руб.


### Функция для расчета рисков и прибыли для каждого региона:
* Применим технику Bootstrap с 1000 выборок, чтобы найти распределение прибыли
* Найдем среднюю прибыль, 95%-й доверительный интервал и риск убытков. 

In [17]:
def risk_calc(target, predictions):
    
    bootstrap_samples = 1000
    alpha = 0.05
    incomes = []
    
    state = np.random.RandomState(0)
    
    for i in range(bootstrap_samples):
        income = pd.Series(predictions).sample(n = TOTAL_WELLS, replace=True, random_state=state)
        incomes.append(predict_income(target, income))

    incomes = pd.Series(incomes)
    income_mean = incomes.mean()
    print('Средняя прибыль: {:.0f}'.format(income_mean), 'руб.')
    
    lower = incomes.quantile(alpha / 2)
    upper = incomes.quantile(1 - alpha / 2)
    
    loss_count = 0
    for inc in incomes:
        if inc < 0:
            loss_count += 1
    loss_rate = loss_count / bootstrap_samples
    print('Риск убытков: {:.2f} %'.format(loss_rate * 100))
    
          
    return print('95% доверительный интервал от {:.2f} до {:.2f} руб.'.format(lower, upper))

### Регион 1

In [18]:
risk_calc(target_valid_0, predictions_0)

Средняя прибыль: 442871777 руб.
Риск убытков: 3.90 %
95% доверительный интервал от -49234991.85 до 921134683.98 руб.


### Регион 2

In [19]:
risk_calc(target_valid_1, predictions_1)

Средняя прибыль: 472926700 руб.
Риск убытков: 0.90 %
95% доверительный интервал от 71697097.80 до 832287542.52 руб.


### Регион 3

In [20]:
risk_calc(target_valid_2, predictions_2)

Средняя прибыль: 384657761 руб.
Риск убытков: 7.20 %
95% доверительный интервал от -134529829.24 до 900128467.31 руб.


### Вывод

Данные были обработаны и рассчитаны: 
- средняя прибыль; 
- 95%-й доверительный интервал;
- риск убытков.

Самый высокий показатель прибыли установлен для второго региона (472 млн. руб.) при самом низком риске убытков (0.9%). Таким образом, для разработки месторождения можно рекомендовать именно этот регион.

## Общий вывод

При выполнении проекта было выполнено:

- Изучены данные. Проведена подготовка данных к моделированию;
- Обучена модель для каждого региона;
- Составлена функция расчета прибыли в регионе;
- Посчитана прибыль и риски за каждый регион и обоснован выбор региона;
- Второй регион получился одним из лучших по показателям предсказывания.